### PR TITLE
Support OpenStack Swift object storage

### PR DESCRIPTION
**Pull request by sudhi-vm**
This commit adds support for OpenStack swift storage.

## Changes

- OpenStack Swift is a highly available, distributed object/blob store. This commit adds support to persist the metric data on Swift.

## Verification

- Ensured `TestObjStore_AcceptanceTest_e2e` works for the swift storage. Verified that the test objects are getting created by going directly to swift while the tests were running.
- Tested with a local OpenStack swift installation.

### CODE REVIEW

**Comment by bwplotka** [comment]
Thanks. Can you add SKIP envvar to `.circleci/config` to make CI green?

File: pkg/objstore/swift/swift.go [hunk 1]
```diff
@@ -0,0 +1,314 @@
+/*
+Copyright 2018 eBay Inc.
```
Comment (domgreen) [inline pkg/objstore/swift/swift.go]: Can we remove the licence comment to keep in line with rest of the repo

File: pkg/objstore/swift/swift.go [hunk 2]
```diff
@@ -0,0 +1,314 @@
...
+func (*Container) Close() error {
+	// nothing to close
+	return nil
+}
+
+func (c *Container) CreateContainer(name string) error {
+	return containers.Create(c.client, name, nil).Err
+}
+
+func (c *Container) DeleteContainer(name string) error {
```
Comment (domgreen) [inline pkg/objstore/swift/swift.go]: Does `Delete|Create` container need to be exported?

File: pkg/objstore/swift/swift.go [hunk 3]
```diff
@@ -0,0 +1,314 @@
...
+		conf.Username == "" ||
+		conf.Password == "" ||
+		(conf.TenantName == "" && conf.TenantID == "") ||
+		conf.RegionName == "" {
+		return errors.New("insufficient swift test configuration information")
+	}
+	return nil
+}
+
+// NewTestContainer creates test bkt client that before returning creates temporary container.
```
Comment (domgreen) [inline pkg/objstore/swift/swift.go]: `bkt` > `objStore`?

File: pkg/objstore/swift/swift.go [hunk 4]
```diff
@@ -0,0 +1,314 @@
...
+		TenantName:    os.Getenv("OS_TENANT_NAME"),
+		RegionName:    os.Getenv("OS_REGION_NAME"),
+		ContainerName: os.Getenv("OS_CONTAINER_NAME"),
+	}
+
+	return c
+}
+
+// ValidateForTests checks to see the config options for tests are set.
+func ValidateForTests(conf swiftConfig) error {
```
Comment (domgreen) [inline pkg/objstore/swift/swift.go]: As above doesnt look like it needs to be exported.

File: pkg/objstore/swift/swift.go [hunk 5]
```diff
@@ -0,0 +1,314 @@
...
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	c := &Container{
+		logger: logger,
+		client: client,
+		opsTotal: prometheus.NewCounterVec(prometheus.CounterOpts{
+			Name:        "thanos_objstore_swift_container_operations_total",
```
Comment (domgreen) [inline pkg/objstore/swift/swift.go]: Not sure we need this here as the factory then uses `BucketWithMetrics`

Comment (domgreen) [reply pkg/objstore/swift/swift.go]: Means we can also remove the `Inc()` calls

File: pkg/objstore/swift/swift.go [hunk 6]
```diff
@@ -0,0 +1,314 @@
...
+	AuthUrl       string `yaml:"authUrl"`
+	Username      string `yaml:"username,omitempty"`
+	UserId        string `yaml:"userId,omitempty"`
+	Password      string `yaml:"password"`
+	DomainId      string `yaml:"domainId,omitempty"`
+	DomainName    string `yaml:"domainName,omitempty"`
+	TenantID      string `yaml:"tenantId,omitempty"`
+	TenantName    string `yaml:"tenantName,omitempty"`
+	RegionName    string `yaml:"regionName,omitempty"`
+	ContainerName string `yaml:"containerName"`
```
Comment (domgreen) [inline pkg/objstore/swift/swift.go]: We need to standardise on the yaml casing ... think we should stick with what Prom is currently using so `snake_case` see new issue #596 .
@bwplotka thoughts?

File: pkg/objstore/objtesting/foreach.go [hunk 7]
```diff
@@ -66,4 +67,21 @@ func ForeachStore(t *testing.T, testFn func(t testing.TB, bkt objstore.Bucket))
...
 		t.Log("THANOS_SKIP_S3_AWS_TESTS envvar present. Skipping test against S3 AWS.")
 	}
+
+	// Optional SWIFT.
+	if _, ok := os.LookupEnv("THANOS_SKIP_SWIFT_TESTS"); !ok {
+		container, closeFn, err := swift.NewTestContainer(t)
+		testutil.Ok(t, err)
+
+		ok := t.Run("swift", func(t *testing.T) {
+			// TODO(bplotka): Add leaktest when https://github.com/GoogleCloudPlatform/google-cloud-go/issues/1025 is resolved.
```
[the remark this answers is not in the document]
Comment (sudhi-vm) [reply pkg/objstore/objtesting/foreach.go]: Removed comment.

File: pkg/objstore/swift/swift.go (hunk 1 above)
Comment (sudhi-vm) [reply pkg/objstore/swift/swift.go]: Removed.

File: pkg/objstore/swift/swift.go (hunk 2 above)
Comment (sudhi-vm) [reply pkg/objstore/swift/swift.go]: updated.

File: pkg/objstore/swift/swift.go (hunk 3 above)
Comment (sudhi-vm) [reply pkg/objstore/swift/swift.go]: updated.

File: pkg/objstore/swift/swift.go (hunk 4 above)
Comment (sudhi-vm) [reply pkg/objstore/swift/swift.go]: updated.

File: pkg/objstore/swift/swift.go (hunk 5 above)
Comment (sudhi-vm) [reply pkg/objstore/swift/swift.go]: removed.

File: pkg/objstore/swift/swift.go (hunk 6 above)
Comment (sudhi-vm) [reply pkg/objstore/swift/swift.go]: changed to `snake_case`

File: pkg/objstore/swift/swift.go [hunk 8]
```diff
@@ -0,0 +1,269 @@
...
+
+// IsObjNotFoundErr returns true if error means that object is not found. Relevant to Get operations.
+func (c *Container) IsObjNotFoundErr(err error) bool {
+	_, ok := err.(gophercloud.ErrDefault404)
+	return ok
+}
+
+// Upload writes the contents of the reader as an object into the container.
+func (c *Container) Upload(ctx context.Context, name string, r io.Reader) error {
+
```
Comment (domgreen) [inline pkg/objstore/swift/swift.go]: nit - blank line

File: pkg/objstore/swift/swift.go [hunk 9]
```diff
@@ -0,0 +1,269 @@
...
+	options := &objects.CreateOpts{Content: r}
+	res := objects.Create(c.client, c.name, name, options)
+	return res.Err
+}
+
+// Delete removes the object with the given name.
+func (c *Container) Delete(ctx context.Context, name string) error {
+	return objects.Delete(c.client, c.name, name, nil).Err
+
```
Comment (domgreen) [inline pkg/objstore/swift/swift.go]: nit - blank line

File: pkg/objstore/swift/swift.go (hunk 8 above)
Comment (sudhi-vm) [reply pkg/objstore/swift/swift.go]: fixed

File: pkg/objstore/swift/swift.go (hunk 9 above)
Comment (sudhi-vm) [reply pkg/objstore/swift/swift.go]: fixed

**Comment by sudhi-vm** [comment]
Hi @domgreen, took care of blank lines and rebased.

File: docs/storage.md [hunk 10]
```diff
@@ -184,4 +185,22 @@ config:
...
+
+```yaml
+type: SWIFT
+config:
+    authUrl: <identity endpoint aka auth URL>
+    username: <username>
+    password: <password>
+    tenantName: <tenantName>
+    regionName: <region>
+    containerName: <container>
```
Comment (domgreen) [inline docs/storage.md]: can we update this to `snake_case`

Comment (sudhi-vm) [reply docs/storage.md]: @domgreen sorry, missed this one in the last commit. fixed it now.

**Comment by sudhi-vm** [comment]
@domgreen This has been running fine in production internally for a few weeks now :)

File: pkg/objstore/objtesting/acceptance_e2e_test.go [hunk 11]
```diff
@@ -62,7 +63,10 @@ func TestObjStore_AcceptanceTest_e2e(t *testing.T) {
 			seen = append(seen, fn)
 			return nil
 		}))
-		testutil.Equals(t, []string{"obj_5.some", "id1/", "id2/"}, seen)
+		expected := []string{"obj_5.some", "id1/", "id2/"}
+		sort.Strings(expected)
+		sort.Strings(seen)
```
Comment (bwplotka) [inline pkg/objstore/objtesting/acceptance_e2e_test.go]: This changes the interface invariants (: I don't think we rely on this though, so "should be" fine. Have you expierience SWIFT iterating in random order? Or just differently ordered?

File: pkg/objstore/swift/swift.go [hunk 12]
```diff
@@ -0,0 +1,267 @@
...
+	}
+
+	client, err := openstack.NewObjectStorageV1(provider, gophercloud.EndpointOpts{
+		Region: sc.RegionName,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	c := &Container{
```
Comment (bwplotka) [inline pkg/objstore/swift/swift.go]: You could potentially inline this, but not a blocker.

File: pkg/objstore/swift/swift.go [hunk 13]
```diff
@@ -0,0 +1,267 @@
...
+// object name including the prefix of the inspected directory.
+func (c *Container) Iter(ctx context.Context, dir string, f func(string) error) error {
+	// Ensure the object name actually ends with a dir suffix. Otherwise we'll just iterate the
+	// object itself as one prefix item.
+	if dir != "" {
+		dir = strings.TrimSuffix(dir, DirDelim) + DirDelim
+	}
+
+	options := &objects.ListOpts{Full: false, Prefix: dir, Delimiter: DirDelim}
+	err := objects.List(c.client, c.name, options).EachPage(func(page pagination.Page) (bool, error) {
```
Comment (bwplotka) [inline pkg/objstore/swift/swift.go]: Just return this. `err` variable is not needed.

File: pkg/objstore/swift/swift.go [hunk 14]
```diff
@@ -0,0 +1,267 @@
...
+		return true, nil
+	})
+
+	return err
+}
+
+// Get returns a reader for the given object name.
+func (c *Container) Get(ctx context.Context, name string) (io.ReadCloser, error) {
+	if name == "" {
+		return nil, fmt.Errorf("error, empty container name passed")
```
Comment (bwplotka) [inline pkg/objstore/swift/swift.go]: Consistent way is to do  `errors.New`   here. (or `errors.Errorf` if you have argument to `f` but seems not)

File: pkg/objstore/swift/swift.go [hunk 15]
```diff
@@ -0,0 +1,267 @@
...
+		Newest: true,
+		Range:  fmt.Sprintf("bytes=%d-%d", off, off+length-1),
+	}
+	response := objects.Download(c.client, c.name, name, options)
+	return response.Body, response.Err
+}
+
+// Exists checks if the given object exists.
+func (c *Container) Exists(ctx context.Context, name string) (bool, error) {
+	err := objects.Get(c.client, c.name, name, nil).Err
```
Comment (bwplotka) [inline pkg/objstore/swift/swift.go]: what `Get` returns? Pointer or value?

Comment (bwplotka) [reply pkg/objstore/swift/swift.go]: Ok, value - all good then (:

File: pkg/objstore/swift/swift.go [hunk 16]
```diff
@@ -0,0 +1,267 @@
...
+	return containers.Create(c.client, name, nil).Err
+}
+
+func (c *Container) deleteContainer(name string) error {
+	return containers.Delete(c.client, name).Err
+}
+
+func configFromEnv() swiftConfig {
+	c := swiftConfig{
+		AuthUrl:       os.Getenv("OS_AUTH_URL"),
```
Comment (bwplotka) [inline pkg/objstore/swift/swift.go]: Not a blocker but wonder if we are sticking to SWIFT name everywhere else, might be better to do SWIFT prefix here as well? Instead OS which can mean lots of things (:

File: pkg/objstore/objtesting/acceptance_e2e_test.go (hunk 11 above)
Comment (sudhi-vm) [reply pkg/objstore/objtesting/acceptance_e2e_test.go]: SWIFT iteration is deterministic, that is, it returns in the same order each time it is queried. It is just differently ordered, which causes the test to fail.

File: pkg/objstore/swift/swift.go (hunk 12 above)
Comment (sudhi-vm) [reply pkg/objstore/swift/swift.go]: done

File: pkg/objstore/swift/swift.go (hunk 13 above)
Comment (sudhi-vm) [reply pkg/objstore/swift/swift.go]: done

File: pkg/objstore/swift/swift.go (hunk 14 above)
Comment (sudhi-vm) [reply pkg/objstore/swift/swift.go]: done

File: pkg/objstore/swift/swift.go (hunk 16 above)
Comment (sudhi-vm) [reply pkg/objstore/swift/swift.go]: I thought about the same... In OpenStack, the params are prefixed with `OS_` (OS=OpenStack), although the acronym is overloaded :) The OpenStack community is familiar with this notation, also CLIs & hundreds of tools typically use this notation.

We can change this, but it's one of those conventions, I think would make it easy for OpenStack folks to easily understand the params when using Thanos with Swift.

**Comment by sudhi-vm** [comment]
Hi @bwplotka, addressed the review comments, please see above.

File: pkg/objstore/swift/swift.go (hunk 16 above)
Comment (bwplotka) [reply pkg/objstore/swift/swift.go]: Ah, yea i remember this now. This makes sense totally.

**Comment by bwplotka** [comment]
Nice work, Thanks!

**Pull request merged**